File: extractPdfText.ts
import { drive } from './src/google/googleClient.js';
import fs from 'fs';
import { Readable } from 'stream';
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const pdfParseModule = require('pdf-parse');
const pdfParse = pdfParseModule.PDFParse || pdfParseModule;

async function downloadPdf(fileId: string): Promise<Buffer> {
  const response = await drive.files.get(
    { fileId, alt: 'media' },
    { responseType: 'stream' }
  );

  const chunks: Buffer[] = [];
  return new Promise((resolve, reject) => {
    (response.data as Readable).on('data', (chunk: Buffer) => {
      chunks.push(chunk);
    });
    (response.data as Readable).on('end', () => {
      resolve(Buffer.concat(chunks));
    });
    (response.data as Readable).on('error', reject);
  });
}

async function extractTextFromPdf(pdfBuffer: Buffer, maxLines: number = 10): Promise<string[]> {
  const PDFParse = pdfParseModule.PDFParse;
  const parser = new PDFParse({ data: pdfBuffer });
  const result = await parser.getText();
  const lines = result.text
    .split('\n')
    .map((line: string) => line.trim())
    .filter((line: string) => line.length > 0);
  return lines.slice(0, maxLines);
}

async function main() {
  const fileId = '1rZJPYvG4QmXXVMkhUmfaSRX3kS24lc64';
  const maxLines = 10;

  try {
    console.log('Downloading PDF from Google Drive...');
    const pdfBuffer = await downloadPdf(fileId);
    console.log(`Downloaded ${pdfBuffer.length} bytes\n`);

    console.log('Extracting text from PDF...');
    const lines = await extractTextFromPdf(pdfBuffer, maxLines);

    console.log(`\n${'='.repeat(60)}`);
    console.log(`First ${lines.length} lines of 'Internation cyber laws.pdf':`);
    console.log('='.repeat(60) + '\n');

    lines.forEach((line, index) => {
      console.log(`${String(index + 1).padStart(2)}. ${line}`);
    });

    console.log(`\n${'='.repeat(60)}`);
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

main();

File: extract_pdf_text.py
#!/usr/bin/env python3
"""
Script to extract text from a PDF file stored in Google Drive.
Downloads the PDF and extracts the first N lines of text.
"""

import sys
import urllib.request
from io import BytesIO

try:
    import PyPDF2
except ImportError:
    print("PyPDF2 is not installed. Installing...")
    import subprocess
    try:
        subprocess.check_call([sys.executable, "-m", "pip", "install", "--user", "PyPDF2"], 
                            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        import PyPDF2
    except:
        print("Error: Could not install PyPDF2. Please install it manually:")
        print("  python3 -m pip install --user PyPDF2")
        sys.exit(1)


def download_pdf_from_drive(file_id: str) -> bytes:
    """Download PDF from Google Drive using the file ID."""
    download_url = f"https://drive.google.com/uc?id={file_id}&export=download"
    
    print(f"Downloading PDF from Google Drive...")
    try:
        with urllib.request.urlopen(download_url) as response:
            return response.read()
    except urllib.error.HTTPError as e:
        raise Exception(f"HTTP Error {e.code}: {e.reason}")
    except urllib.error.URLError as e:
        raise Exception(f"URL Error: {e.reason}")


def extract_text_from_pdf(pdf_bytes: bytes, max_lines: int = 10) -> list[str]:
    """Extract text from PDF bytes and return first N lines."""
    pdf_file = BytesIO(pdf_bytes)
    pdf_reader = PyPDF2.PdfReader(pdf_file)
    
    all_text = []
    
    print(f"Extracting text from {len(pdf_reader.pages)} pages...")
    
    # Extract text from all pages
    for page_num, page in enumerate(pdf_reader.pages, 1):
        try:
            text = page.extract_text()
            if text:
                lines = text.split('\n')
                all_text.extend([line.strip() for line in lines if line.strip()])
        except Exception as e:
            print(f"Warning: Could not extract text from page {page_num}: {e}")
    
    # Return first max_lines
    return all_text[:max_lines]


def main():
    # Google Drive file ID for "Internation cyber laws.pdf"
    file_id = "1rZJPYvG4QmXXVMkhUmfaSRX3kS24lc64"
    max_lines = 10
    
    try:
        # Download PDF
        pdf_bytes = download_pdf_from_drive(file_id)
        print(f"Downloaded {len(pdf_bytes)} bytes\n")
        
        # Extract text
        lines = extract_text_from_pdf(pdf_bytes, max_lines)
        
        # Display results
        print(f"\n{'='*60}")
        print(f"First {len(lines)} lines of 'Internation cyber laws.pdf':")
        print(f"{'='*60}\n")
        
        for i, line in enumerate(lines, 1):
            print(f"{i:2d}. {line}")
        
        print(f"\n{'='*60}")
        
    except Exception as e:
        print(f"Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()

